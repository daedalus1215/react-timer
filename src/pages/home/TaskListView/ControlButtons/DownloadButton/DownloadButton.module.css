
.buttonDownload,
.buttonDownload:hover {
  display: flex;
  background-color: #3e80ed;
  color: white;
  border-radius: 4px;
  width: 90px;
  height: 1.5em;
  border: none;
  padding: 1em;
  text-align: center;
  text-decoration: none;
  margin-left: 0.5em;
}

.buttonDownload:hover {
  background-color: #184693;
}