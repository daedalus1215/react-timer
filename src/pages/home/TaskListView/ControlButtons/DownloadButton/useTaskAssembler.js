import { getFormattedDate } from 'utils';
import { useFetchProjectOptions } from 'hooks';
import formatTimeContractAndCustomer from './formatTimeContractAndCustomer';

//@TODO: Write a test for this
const useTaskAssembler = (...tasks) => {
    const projects = useFetchProjectOptions();

    const dateFormatted = getFormattedDate(new Date());

    const timeTask = {
        date: dateFormatted
    };

    timeTask.WorkUnit = tasks.map(task =>
        formatTimeContractAndCustomer(task, projects)
    );

    return () => timeTask;
}

export default useTaskAssembler;