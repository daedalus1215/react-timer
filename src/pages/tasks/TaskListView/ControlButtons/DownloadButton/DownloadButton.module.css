.buttonDownload,
.buttonDownload:hover {
  background-color: #3e80ed;
  color: white;
  width: 45px;
  height: 40px;
  margin: 4px 5px;
}

.buttonDownload:hover {
  background-color: #184693;
}
