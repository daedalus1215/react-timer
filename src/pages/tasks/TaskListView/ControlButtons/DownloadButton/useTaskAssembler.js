import { getCurrentDateTimeEstFormat } from 'utils';
import { useFetchProjectOptions } from 'hooks';
import formatTimeContractAndCustomer from './formatTimeContractAndCustomer';

//@TODO: Write a test for this
const useTaskAssembler = (...tasks) => {
    const projects = useFetchProjectOptions();

    const dateFormatted = getCurrentDateTimeEstFormat();

    const timeTask = {
        date: dateFormatted
    };

    timeTask.WorkUnit = tasks?.map(task =>
        formatTimeContractAndCustomer(task, projects)
    );

    return () => timeTask;
}

export default useTaskAssembler;