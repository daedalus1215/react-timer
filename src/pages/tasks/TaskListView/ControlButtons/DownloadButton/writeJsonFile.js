var FileSaver = require('file-saver');

/**
 * Not exporting default, so we can spyOn
 * @param {Object} taskBundle
 */
export function writeJsonFile(taskBundle) {
  const writeable = {
    date: taskBundle.date,
    WorkUnit: taskBundle.WorkUnit[0][0].tasks,
  }
  window.console.log('writeable', writeable)
  let json = JSON.stringify(writeable);
  let blob = new Blob([json], { type: 'application/json' });
  let fileName = 'time-logs_' + taskBundle.date + '.json';
  FileSaver.saveAs(blob, fileName);
}
