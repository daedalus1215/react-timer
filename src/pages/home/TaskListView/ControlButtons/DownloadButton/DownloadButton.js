import React from 'react';
import { useTaskEditContext } from 'hooks';
import { writeJsonFile } from './writeJsonFile';
import useTaskAssembler from './useTaskAssembler';
import styles from './DownloadButton.module.css';

const DownloadButton = tasks => {
    const assembleTask = useTaskAssembler(tasks);

    const handleDownload = () => {
        writeJsonFile(assembleTask());
    };

    return (
        <>
            {tasks.length || (
                <button
                    type="a"
                    className={styles.buttonDownload}
                    onClick={handleDownload}
                    data-test-id="btn-download">
                    <span className="glyphicon glyphicon-download-alt mr-5px"></span>
                </button>)
            }
        </>);
};

export default DownloadButton;