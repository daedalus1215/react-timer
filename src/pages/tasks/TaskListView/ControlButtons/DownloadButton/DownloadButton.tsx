import React from 'react';
import cn from 'classnames';
import { Task } from 'interfaces/pages/tasks/Task';
import { Button } from 'components';
import { writeJsonFile } from './writeJsonFile';
import useTaskAssembler from './useTaskAssembler';
import styles from './DownloadButton.module.css';

interface DownloadButtonProps {
  tasks: Task[];
}

const useHandleDownload = (tasks: Task[]) => {
  const assembleTask = useTaskAssembler(tasks);
  return () => {
    writeJsonFile(assembleTask());
  };
};

const DownloadButton: React.FC<DownloadButtonProps> = ({ tasks }) => {
  const handleDownload = useHandleDownload(tasks);

  return <>{tasks.length || <Button title="Download Tasks" type="a" className={cn(styles.buttonDownload, 'glyphicon glyphicon-download-alt')} onClick={handleDownload} />}</>;
};

export default DownloadButton;
