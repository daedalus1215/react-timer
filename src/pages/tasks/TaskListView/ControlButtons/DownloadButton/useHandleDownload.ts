import { writeJsonFile } from './writeJsonFile';
import useTaskAssembler from './useTaskAssembler';
import { TaskInterface } from 'interfaces/pages/tasks/Task';

const useHandleDownload = (tasks: TaskInterface[]) => {
    const assembleTask = useTaskAssembler(tasks);
    return () => {
        writeJsonFile(assembleTask());
    };
};

export default useHandleDownload;