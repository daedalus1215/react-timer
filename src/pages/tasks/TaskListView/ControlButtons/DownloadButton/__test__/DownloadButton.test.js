import React from 'react';
import { render } from '@testing-library/react';
import DownloadButton from '../DownloadButton';

describe('src/pages/tasks/TaskListView/__test__/DownloadButton.test.js', () => {
    describe('DownloadButton', () => {
        it('should render the button when tasks are available', () => {
            // Arrange
            const tasks = [{ _id: '1', }, { _id: 'yup' }];

            // Act
            const { queryByTestId } = render(<DownloadButton tasks={tasks} />);

            // Assert
            expect(queryByTestId('btn-download')).toBeInTheDocument();
            expect(queryByTestId('btn-download')).toMatchSnapshot();
        });

        it('should not render the button when tasks are not available', () => {
            // Arrange
            const tasks = [];

            // Act
            const { queryByTestId } = render(<DownloadButton tasks={tasks} />);

            // Assert
            expect(queryByTestId('btn-delete')).not.toBeInTheDocument();
        });
    });
});