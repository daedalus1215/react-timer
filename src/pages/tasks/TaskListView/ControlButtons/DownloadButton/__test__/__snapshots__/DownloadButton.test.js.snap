// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/pages/tasks/TaskListView/__test__/DownloadButton.test.js DownloadButton should render the button when tasks are available 1`] = `
<button
  class="btn btn-default navbar-btn baseBtn buttonDownload glyphicon glyphicon-download-alt"
  data-testid="btn-download"
/>
`;
